drivers/emul/eeprom_at24_emul.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_at24
+
+#define LOG_LEVEL CONFIG_I2C_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(atmel_at24);
+
+#include <device.h>
+#include <drivers/i2c.h>
+#include <drivers/i2c_emul.h>
+
+#define REG_COUNT	0x100
+
+/** Defines run-time data used by the emulator */
+struct at24_emul_data {
+	/** Size of EEPROM */
+	uint16_t size;
+	/** EEPROM data contents */
+	uint8_t *buf;
+	/** Current register to read */
+	uint8_t cur_reg;
+	/** Current I2C config */
+	uint32_t dev_config;
+	/** Address width for EEPROM in bits (8 or 16) */
+	uint8_t addr_width;
+	/** true if the bus is too fast */
```

Comment:
This allows the emulator to fail if the speed is too high for what it expects, so that we can test that behaviour. This is just an example of the sorts of things that can be done.